[오렌지] 로또1단계 미션 제출합니다

YerinCho opened the pull request:
리뷰 요청 합니다. 감사합니다!

YerinCho committed:
[Lotto] docs : 구현 목록 작성

YerinCho committed:
[Lotto] feat : LottoFactory

 로또 번호 생성 기능 추가
 무작위로 6개 숫자 선택해 로또 생성

YerinCho committed:
[Lotto] refactor : LottoFactory

 상수화, 불필요한 코드 삭제

YerinCho committed:
[Lotto] docs : 구현 목록 수정

YerinCho committed:
[Lotto] feat : PurchaseLotto

 로또 구매 갯수 계산

YerinCho committed:
[Lotto] refactor : PurchaseLotto

 상수화

YerinCho committed:
[Lotto] docs : 구현목록 수정

YerinCho committed:
[Lotto] refactor : LottoFactoryTest

 불필요한 코드 제거

YerinCho committed:
[Lotto] feat : Lotto, 중복 검사 구현

YerinCho committed:
[Lotto] feat : InputView

YerinCho committed:
[Lotto] docs : InputView

YerinCho committed:
[Lotto] feat : InputView 입력 예외처리

YerinCho committed:
[Lotto] feat : PurchaseLotto 입력 예외처리

 입력한 금액이 1000원 미만이면 예외처리

YerinCho committed:
[Lotto] docs : README 수정

 구입 금액 입력 예외처리
 보너스볼 입력 예외처리

YerinCho committed:
[Lotto] feat : Buyer 생성자 구현

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : Lotto.validateSize

 6자리인지 검사

YerinCho committed:
[Lotto] feat : SplitLottoNumbers

 , 기준으로 String 분리

YerinCho committed:
[Lotto] feat : Lotto.validateNumberScope

 로또 번호의 범위 유효성 검사 추가

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : WinningLotto

 보너스볼 범위 유효성 검사 추가
 Lotto.validateNumbersScope 분리

YerinCho committed:
[Lotto] feat : WinningLotto 보너스볼 중복검사

 생성자 수정
 Lotto.lottoNumbers 접근제어자 변경

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : WinnigResult 로또 일치 갯수 계산

YerinCho committed:
[Lotto] feat : WinningRankResult 보너스볼 일치 여부 확인

 WinnigResult -> WinningRankResult

YerinCho committed:
[Lotto] feat : WinningRankResult, WinningValue

 각 매수 별 당첨된 등수를 확인하는 기능 추가
 Lotto.getLottoNumbers 추가
 WinningLotto.isBonusNumber 추가
 WinningRankResult.isBonusNumberContain, WinningRankResult.checkCorrectNumber 의 파라미터 변경

YerinCho committed:
[Lotto] feat : RewardRate

 수익률 계산 기능 추가
 WinningRankResult.getWinningValueResult 추가

YerinCho committed:
[Lotto] refactor : RewardRate

 for 문 stream 으로 변경

YerinCho committed:
[Lotto] refactor : WinningRankResult.checkRank 메소드 분리

YerinCho committed:
[Lotto] refactor : Lotto.getLottoNumbers 불변성 보장

YerinCho committed:
`[Lotto] refactor : PurchaseLotto -> MoneyManager 클래스명 변경

 생성자 추가, 멤버변수 money 추가

YerinCho committed:
[Lotto] refactor : Buyer 멤버변수 제거

YerinCho committed:
[Lotto] fix : InputView 입력 버퍼 초기화

YerinCho committed:
[Lotto] feat : Buyer.getLottos 추가

YerinCho committed:
[Lotto] feat : 초기 구매 화면 구현

 LottoMain 추가
 LottoManager 추가
 OutputView 추가

YerinCho committed:
[Lotto] feat : 당첨번호, 보너스번호 입력 구현

 LottoManager.setWinningLotto 추가

YerinCho committed:
[Lotto] feat : 당첨 결과 화면 구현

 LottoManager.analyzeLotto 추가
 WinningRankResult.winningValueResult 자료형 LinkedHashMap 으로 변경
 WinningValue 메시지 추가

YerinCho committed:
[Lotto] fix : 입력 버퍼 오류 수정

YerinCho committed:
[Lotto] feat : 수익률 계산 결과 출력 구현

 수익률 계산식 수정
 LottoManager.analyzeRewardRate 추가
 MoneyManager.getMoney 추가

YerinCho committed:
[Lotto] feat : 예외 메시지 추가

YerinCho committed:
[Lotto] refactor

 메소드, 변수명, 클래스명 변경
 메소드 분리
 접근 제한자 변경
 원시값 상수화
 메소드가 위치하는 클래스 변경
 불필요한 코드 정리
 README 수정

YerinCho committed:
[Lotto] feat : LottoNumber,ConvertInput 추가

 Integer 형식의 로또 넘버를 LottoNumber 로 포장
 그에 따른 ConvertInput 수정

YerinCho committed:
[Lotto] refactor : 패키지 분리

YerinCho committed:
[Lotto] feat : ConvertInputTest 공백 제거 검사

YerinCho committed:
[Lotto] feat : LottoNumber

 로또 숫자 범위 확인 테스트

YerinCho committed:
[Lotto] refactor : ConvertInput

 문자열 상수화

YerinCho committed:
[Lotto] refactor : LottoFactory

 singleton 패턴으로 변경

jeonghoon1107 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Map;
+
+public class LottoManager {
+    public Money money;
```
기본적으로 인스턴스 변수가 public이면 예기치 못한 많은 예외상황이 발생할 수 있습니다.
해당 부분에 대해 공부해보시면 좋을 것 같아요.

YerinCho replied:
OutputView 로 인자를 보내게 되어서 public 으로 했던 것이었는데요. private 으로 바꾸고 getter를 사용했습니다. 궁금한 점이 있는데, LottoManager 의 money 와 buyer를 없애고, main 의 로컬 변수로 선언하는 방법과 LottoManager에서 getter 를 사용하는 방법 중 어느 것이 나은 지 궁금합니다. main 에서 선언하면 마찬가지로 외부의 변경에 취약할 것 같아서요.

jeonghoon1107 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+import java.util.Map;
+
+public class LottoManager {
```
LottoManager가 하는 일이 너무 많은데 분리하면 어떨까요?

YerinCho replied:
분석 용도의 두 메서드를 LottoAnalyzeManager 클래스로 분리했습니다.
그런데 분리를 하다 보니 getter를 많이 쓰게 되네요 ㅠㅠ

jeonghoon1107 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,27 @@
+        this.money = money;
+    }
+
+    static void validateMoney(int money) {
```
static 이어야 하나요?

YerinCho replied:
테스트 때문에 그렇게 했었던 것인데, 생각해 보니 테스트는 생성자를 통해 하면 될 것 같네요. static 제거하고 테스트 수정 했습니다!

jeonghoon1107 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,27 @@
+package lotto.domain;
+
+public class Money {
+    private static final int LOTTO_PRICE = 1000;
```
_를 이용하면 가독성을 높일 수 있어요!

jeonghoon1107 commented on src/main/java/lotto/domain/result/WinningValue.java:
```diff
@@ -0,0 +1,39 @@
+    FORTH(4, 50000, "4개 일치"),
+    THIRD(5, 1500000, "5개 일치"),
+    SECOND(5, 30000000, "5개 일치, 보너스번호 일치"),
+    FIRST(6, 2000000000, "6개 일치");
```
bonus 여부를 함께 들고 있으면 어떨까요?

YerinCho replied:
기존에 있던 if 문을 사용할 필요가 없어지네요!

jeonghoon1107 commented on src/main/java/lotto/domain/result/WinningValue.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static WinningValue valueOf(int hitCount, boolean hitBonus) {
+        if (hitCount == SECOND.hitCount && hitBonus) return SECOND;
```
if문에 나열되는 조건들이 많아질수록 가독성이 떨어지는데 메서드로 추출하면 명확하게 할 수 있을 것 같아요.

YerinCho replied:
2등인지 확인하는 부분이므로 isSecond 메서드로 분리했습니다!

jeonghoon1107 commented on src/main/java/lotto/domain/lottoTicket/WinningLotto.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.List;
+
+public class WinningLotto extends Lotto {
```
상속과 조합의 장단점에 대해 공부해보면 좋을 것 같아요.

jeonghoon1107 commented on src/main/java/lotto/domain/lottoTicket/LottoNumber.java:
```diff
@@ -0,0 +1,39 @@
+
+    private int lottoNumber;
+
+    public LottoNumber(int lottoNumber) {
```
LottoNumber를 어떻게 재사용할 수 있을까요?

YerinCho replied:
WinningLotto 의 보너스 넘버를 LottoNumber로 선언하여 사용하고 있긴 한데, 그 외 재사용할 수 있는 부분은 잘 모르겠습니다. ㅠㅠ

jeonghoon1107 reviewed (changes requested):
안녕하세요 오렌지!
피드백 남겼으니 확인 부탁드려요 :)

YerinCho committed:
[Lotto] refactor : LottoManager

 LottoManager 의 인스턴스 변수 Buyer, Money 제거
 main 으로 위치 변경

YerinCho committed:
[Lotto] refactor : LottoManager

 인스턴스 변수의 접근제어자 변경, getter 추가

YerinCho committed:
[Lotto] refactor : Money

 validateMoney 의 static 제거 및 그에 따른 테스트 변경

YerinCho committed:
[Lotto] refactor : LottoNumber

 validateNumberScope 의 static 제거 및 그에 따른 테스트 변경

YerinCho committed:
[Lotto] refactor : WinningValue, Money

 숫자에 _추가

YerinCho committed:
[Lotto] refactor : WinningValue

 메서드 분리

YerinCho committed:
[Lotto] refactor : LottoResult

 수익률 식 변경

YerinCho committed:
[Lotto] refactor : WinningLotto

 Lotto 를 상속하지 않고, Lotto WinningLotto 의 형식 사용으로 변경

YerinCho committed:
[Lotto] refactor : LottoResultTest

 수익률 식 변경

YerinCho committed:
[Lotto] refactor : LottoManager, LottoAnalyzeManager

 LottoManager 에서 로또 결과 관련 메서드들을 LottoAnalyzeManager class 로 분리

YerinCho committed:
[Lotto] refactor : WinnigLotto

 보너스번호 관련 정보 추가

YerinCho committed:
[Lotto] refactor : WinnigLotto

 메서드 분리

YerinCho committed:
[Lotto] fix : LottoResult

 Int 오버플로우로 인한 식 수정

YerinCho committed:
[Lotto] fix : LottoResult

 Int 오버플로우로 인한 식 수정

YerinCho committed:
[Lotto] refactor : WinningLotto

 isBonusNumber 에서, getLottoNumber를 사용하는 대신 Lotto에 isContainBonusNumber 메서드를 만들어 사용

jeonghoon1107 commented on src/main/java/lotto/domain/result/LottoResult.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public Map<WinningValue, Integer> getLottoResult() {
+        lottoResult.remove(WinningValue.ZERO);
```
원본 객체를 손상시키지 않으면 좋겠습니다!

jeonghoon1107 commented on src/main/java/lotto/LottoMain.java:
```diff
@@ -0,0 +1,21 @@
+
+        lottoManager.setWinningLotto(InputView.inputWinningLottoNumbers(), InputView.inputBonusNumber());
+
+        LottoAnalyzeManager lottoAnalyzeManager = new LottoAnalyzeManager(lottoManager.getBuyer(), lottoManager.getWinningLotto());
```
LottoAnalyzeManager를 통하지 않고 LottoResult 객체를 생성 하면 어떨까요?

jeonghoon1107 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public void setWinningLotto(String numbers, int bonusNumber) {
+        List<LottoNumber> winningLottoNumbers = ConvertInput.convertLottoNumbers(numbers);
```
ConvertInput을 거치지 않고 new Lotto를 해서 winningLotto에게 넘겨주면 어떨까요?
Lotto를 만들어주는 클래스가 있으면 좋을 것 같아요!

jeonghoon1107 reviewed (approved):
안녕하세요 오렌지!
리뷰가 늦었습니다 ㅠㅠ
몇가지 피드백 남겼으니 다음 미션에 반영해주세요 :)
